tests/Model/LazyBSONDocumentTest.php

```diff
@@ -0,0 +1,339 @@
+    {
+        $document = new LazyBSONDocument(['foo' => 'bar']);
+
+        $this->expectWarning();
+        $this->expectWarningMessage('Undefined offset: 1');
+        $document[1];
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testGetDocument(LazyBSONDocument $document): void
+    {
+        $this->assertInstanceOf(LazyBSONDocument::class, $document->document);
+        $this->assertInstanceOf(LazyBSONDocument::class, $document['document']);
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testGetArray(LazyBSONDocument $document): void
+    {
+        $this->assertInstanceOf(LazyBSONArray::class, $document->array);
+        $this->assertInstanceOf(LazyBSONArray::class, $document['array']);
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testPropertyIsset(LazyBSONDocument $document): void
+    {
+        $this->assertTrue(isset($document->foo));
+        $this->assertFalse(isset($document->bar));
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testOffsetExists(LazyBSONDocument $document): void
+    {
+        $this->assertTrue(isset($document['foo']));
+        $this->assertFalse(isset($document['bar']));
+    }
+
+    public function testOffsetExistsWithInvalidOffset(): void
+    {
+        $document = new LazyBSONDocument(['foo' => 'bar']);
+
+        $this->assertFalse(isset($document[1]));
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testPropertySet(LazyBSONDocument $document): void
+    {
+        $this->assertFalse(isset($document->new));
+        $document->new = 'yay!';
+        $this->assertSame('yay!', $document->new);
+
+        $this->assertSame('bar', $document->foo);
+        $document->foo = 'baz';
+        $this->assertSame('baz', $document->foo);
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testOffsetSet(LazyBSONDocument $document): void
+    {
+        $this->assertFalse(isset($document['new']));
+        $document['new'] = 'yay!';
+        $this->assertSame('yay!', $document['new']);
+
+        $this->assertSame('bar', $document['foo']);
+        $document['foo'] = 'baz';
+        $this->assertSame('baz', $document['foo']);
+    }
+
+    public function testOffsetSetWithInvalidOffset(): void
+    {
+        $document = new LazyBSONDocument(['foo' => 'bar']);
+
+        $this->expectWarning();
+        $this->expectWarningMessage('Unsupported offset: 1');
+        $document[1] = 'foo';
+    }
+
+    /** @dataProvider provideTestDocument */
+    public function testPropertyUnset(LazyBSONDocument $document): void
+    {
+        $this->assertFalse(isset($document->new));
+        $document->new = 'yay!';
+        unset($document->new);
+        $this->assertFalse(isset($document->new));
+
+        unset($document->foo);
+        $this->assertFalse(isset($document->foo));
+
+        // Change value to ensure it is unset for good
```

Comment:
Similar to my comment in LazyBSONArrayTest, I'm curious what "unset for good" means here.